ocrd/ocrd/cli/resmgr.py

```diff
@@ -0,0 +1,114 @@
+    print('%s' % executable)
+    for resdict in reslist:
+        print('- %s (%s)\n  %s' % (resdict['name'], resdict['url'], resdict['description']))
+    print()
+
+@click.group("resmgr")
+def resmgr_cli():
+    """
+    Managing processor resources
+    """
+    initLogging()
+
+@resmgr_cli.command('list-available')
+@click.option('-e', '--executable', help='Show only resources for executable EXEC', metavar='EXEC')
+def list_available(executable=None):
+    """
+    List available resources
+    """
+    resmgr = OcrdResourceManager()
+    for executable, reslist in resmgr.list_available(executable):
+        print_resources(executable, reslist)
+
+@resmgr_cli.command('list-installed')
+@click.option('-e', '--executable', help='Show only resources for executable EXEC', metavar='EXEC')
+def list_installed(executable=None):
+    """
+    List installed resources
+    """
+    resmgr = OcrdResourceManager()
+    ret = []
+    for executable, reslist in resmgr.list_installed(executable):
+        print_resources(executable, reslist)
+
+@resmgr_cli.command('download')
+@click.option('-n', '--any-url', help='Allow downloading unregistered resources', is_flag=True)
+@click.option('-o', '--overwrite', help='Overwrite existing resources', is_flag=True)
+@click.option('-l', '--location', help='Where to store resources', type=click.Choice(['cache', 'config', 'data', 'cwd']), default='cache', show_default=True)
+@click.argument('executable', required=True)
+@click.argument('url_or_name', required=True)
+def download(any_url, overwrite, location, executable, url_or_name):
+    """
+    Download resource URL_OR_NAME for processor EXECUTABLE.
+
+    URL_OR_NAME can either be the ``name`` or ``url`` of a registered resource.
+
+    If URL_OR_NAME is '*' (asterisk), download all known resources for this processor
+
+    If ``--any-url`` is given, also accepts URL of non-registered resources for ``URL_OR_NAME``.
+    """
+    log = getLogger('ocrd.cli.resmgr')
+    resmgr = OcrdResourceManager()
+    basedir = XDG_CACHE_HOME if location == 'cache' else \
+            XDG_DATA_HOME if location == 'data' else \
+            XDG_CONFIG_HOME if location == 'config' else \
+            getcwd()
+    is_url = url_or_name.startswith('https://') or url_or_name.startswith('http://')
+    find_kwargs = {'executable': executable}
+    if url_or_name != '*':
+        find_kwargs['url' if is_url else 'name'] = url_or_name
+    reslist = resmgr.find_resources(**find_kwargs)
+    if not reslist:
+        log.info("No resources found in registry")
+        if is_url and any_url:
+            log.info("Downloading unregistered resource %s" % url_or_name)
+            with requests.head(url_or_name) as r:
+                content_length = int(r.headers.get('content-length'))
+            with click.progressbar(length=content_length) as bar:
+                fpath = resmgr.download(
+                    executable,
+                    url_or_name,
```

Comment:
> IIUC this means the new / unregistered resources will not be visible under list-installed afterwards.

They will, but currently there's just question marks where the description and URL would be for a registered resource:

```
$ touch $HOME/.cache/ocrd-tesserocr-recognize
$ ocrd resmgr list-installed
ocrd-tesserocr-recognize
- bar (???)
  ???
- Fraktur_GT4HistOCR.traineddata (https://ub-backup.bib.uni-mannheim.de/~stweil/ocrd-train/data/Fraktur_5000000/tessdata_fast/Fraktur_50000000.334_450937.traineddata)                       
  Tesseract LSTM model trained on GT4HistOCR
```

You mean, create an entry with URL and name in `$HOME/.config/resources.yml` and create that if it doesn't exist yet? Yes that would work.

> ```ocrd resmgr download -n ocrd-cis-ocropy-recognize https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz```

My reasoning was: If you use a URL once, you will use this URL in the future (and not the short name), hence the mapping from URL to name should be simple to make this reproducible. But I guess "last URL segment sans fragment identifier and query" would work just as well.

Reply:
> You mean, create an entry with URL and name in `$HOME/.config/resources.yml` and create that if it doesn't exist yet? Yes that would work.

Yes, exactly that.

> > `ocrd resmgr download -n ocrd-cis-ocropy-recognize https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz`
> 
> My reasoning was: If you use a URL once, you will use this URL in the future (and not the short name), hence the mapping from URL to name should be simple to make this reproducible. But I guess "last URL segment sans fragment identifier and query" would work just as well.

You mean the user will pass the URL in the parameter JSON? But then why does it say 
> Use in parameters as 'httpsgithubcomzuphilipocropymodelsrawmasterendefaultpyrnngz'

? I would expect either 
> Use in parameters as 'https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz'

or (better)
> Use in parameters as 'en-default.pyrnn.gz'

Reply:
> `Use in parameters as 'httpsgithubcomzuphilipocropymodelsrawmasterendefaultpyrnngz'`

Excellent point, my reasoning was flawed :)

Reply:
BTW, https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz is already in our resource_list.yml, I just had a typo in it, fixed in 54a214a0

Reply:
Maybe (besides `nth_url_segment`) even normalize URL encoding (eg. replace `%C5%BF` by `ſ`)?

Reply:
3ee66ce5 Now:

```
ocrd resmgr download -n ocrd-foo https://github.com/zuphilip/ocropy-french-models/raw/master/models/french-balzac-model-00105000.pyrnn.gz                                                 
15:13:11.315 INFO ocrd.cli.resmgr - No resources found in registry
15:13:11.315 INFO ocrd.cli.resmgr - Downloading unregistered resource https://github.com/zuphilip/ocropy-french-models/raw/master/models/french-balzac-model-00105000.pyrnn.gz               
  [####################################]  100%
15:13:16.579 INFO ocrd.cli.resmgr - Downloaded https://github.com/zuphilip/ocropy-french-models/raw/master/models/french-balzac-model-00105000.pyrnn.gz to /home/kba/.cache/ocrd-foo/french-ba
lzac-model-00105000.pyrnn.gz
15:13:16.579 INFO ocrd.cli.resmgr - Use in parameters as 'french-balzac-model-00105000.pyrnn.gz'

```

Reply:
> Maybe (besides nth_url_segment) even normalize URL encoding (eg. replace %C5%BF by ſ), perhaps even dropping all query/fragment components from the path?

Good point b21d4628 Do you have an example URL to test with?

Reply:
I do – and it _fails_ because the HEAD request does not get answered with a `content-length` header in Gitlab:
- my [s-ſ reconstruction model](https://git.informatik.uni-leipzig.de/ocr-d/cor-asv-ann-models/-/raw/master/s2s.gt4histocr.s-%C5%BF.d2.w0512.adam.attention.stateless.variational-dropout.char.transfer-lm.h5?inline=false)

```
  File "venv/lib/python3.6/site-packages/ocrd/cli/resmgr.py", line 85, in download
    content_length = int(r.headers.get('content-length'))
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

See [here](https://stackoverflow.com/a/44299915/14474237) for a solution (i.e. use GET but with `stream=True`)...

Reply:
Thanks, I will test with that and replace HEAD with bodyless GET. And good to know these resources are mirrorred on your GitLab, I was trying to get support for Git LFS but it is surprisingly tricky to implement without additional dependencies or lots of copy/paste code.

Reply:
d8d97af1 

```
13:38:08.997 INFO ocrd.cli.resmgr - Downloaded https://git.informatik.uni-leipzig.de/ocr-d/cor-asv-ann-models/-/raw/master/s2s.gt4histocr.s-%C5%BF.d2.w0512.adam.attention.stateless.variation
al-dropout.char.transfer-lm.h5?inline=false to /home/kba/.cache/ocrd-foo/s2s.gt4histocr.s-ſ.d2.w0512.adam.attention.stateless.variational-dropout.char.transfer-lm.h5                         
13:38:08.998 INFO ocrd.cli.resmgr - Use in parameters as 's2s.gt4histocr.s-ſ.d2.w0512.adam.attention.stateless.variational-dropout.char.transfer-lm.h5'
```

Reply:
Fantastic!